### ISSUE TITLE
Curious about Glow implementation: some weights look frozen?

### ISSUE THREAD

**Issue by christabella** [kmkolasinski/deep-learning-notes#20]
Hi Krzysztof,

When visualizing the distribution of weights and gradients of each tensor over training, I noticed that some the weights don't seem to be updating; e.g. `InvertibleConv1x1Layer`'s `U_mat`, `L_mat`, and `log_S`.
![screenshot_2019-12-19_11-40-01](https://user-images.githubusercontent.com/2038751/71255264-196cfa80-2336-11ea-9d2e-0f01235bf95d.png)

My first thought was that maybe the gradients are too small, but it doesn't look like that's the case:
![IMAGE 2019-12-20 14:39:48](https://user-images.githubusercontent.com/2038751/71255447-9304e880-2336-11ea-9cd2-feecb1874b27.jpg)

I didn't change the core code and used the high-level API, but trained it on a different task and it is plugged into a larger model.

I will try running the original example you provided and report back with that, but in the meantime I was wondering if you (or anyone else) had any early ideas about this. Thanks!

**Comment by kmkolasinski** [comment]
This is interesting, but unfortunately I have no idea for this. Do you have some minimal example of code to run or test like small model with fake input data ? Are you sure you are plotting the right thing? 
Probably it would be worth to check whether optimizer applies these gradients to L, U, scale weights. Maybe we could start from single layer with 1D-invertible conv and check whether it can learn some expected distribution like gaussian multivariate with nontrivial covariance matrix.

**Comment by christabella** [comment]
Thank you very much for the reply! I just tried visualizing the weights in the provided notebook `Celeba48x48_22steps.ipynb` without changing anything else, and some gradients are also exploding even in step 2:

![image](https://user-images.githubusercontent.com/2038751/71742726-5921ff00-2e6b-11ea-9e4b-1acb596dd757.png)

> Probably it would be worth to check whether optimizer applies these gradients to L, U, scale weights. Maybe we could start from single layer with 1D-invertible conv and check whether it can learn some expected distribution like gaussian multivariate with nontrivial covariance matrix.

If I understand your suggestion correctly, I will try to reduce the flow's complexity (number of steps etc.) and plot the weights again. Maybe on MNIST instead of CelebA...

**Comment by kmkolasinski** [comment]
Yes good idea, maybe it would be better to start from much more smaller learning rate and simpler architecture. MNIST is good idea also. Can you check this ?

**Comment by geosada** [comment]
Though I'm not too sure if it's going to be useful, let me share with you what I encountered while playing with the code for CelebA 64x64.
Increasing flow steps from 22 (the orginal Krzysztof's setting) to 32 caused a loss to get NaN, which I suspect that a gradient explosion probably happened somewhere.

Actually, at that time, I changed the learning dynamics carelessly from the original: I started to run the training with num_steps=1000 and lr_ph: 0.0005 from the beginning, and it turned out that this change caused NaN error after all.

In short, the warm-up strategy, that is, In [33] and In [34] for example in Celeba64x64_22steps.ipynb, seems important.
So, Christabella, I also think that it is worth trying to start the training with much smaller lr_ph, as Krzysztof suggested in the last comment.

**Comment by kmkolasinski** [comment]
@geosada  thanks for your comment :+1:

**Comment by christabella** [comment]
Thanks a lot @geosada for pointing out the importance of the warm-up strategy! Although I have heard of starting with a small learning rate that [goes up and down again](https://bella.cc/zettels/learning_rate_scheduling_warm_up/), but in `Celeba64x64_22steps.ipynb` the learning rate schedule goes up, down, down, up---just wondering, why that schedule?
```
/\    instead of  /\ 
  \/             /  \

[33]: lr=0.0001 for 1 x 100 steps
[34]: lr=0.0005 for 5 x 1000 steps
[35]: lr=0.0001 for 5 x 1000 steps
[36]: lr=0.00005 for 5 x 1000 steps
[37]: lr=0.0001 for 5 x 1000 steps
```

**Comment by christabella** [comment]
I also noticed that the CelebA notebook uses a per-pixel loss
```
loss_per_pixel = loss / image_size / image_size  
total_loss = l2_loss + loss_per_pixel
```
while for MNIST, it's the sum over all pixels
```total_loss = l2_loss + loss```

Maybe that’s why the l2 loss goes up in the MNIST notebook, because the loss summed over all pixels overpowers the l2 loss:
![image](https://user-images.githubusercontent.com/2038751/72153089-8d3e7800-33b5-11ea-8b9a-29ba2a30a679.png)

Furthermore, in the [official GLOW implementation](https://github.com/openai/glow/blob/d3906b2f61183e2178feeb1d63dd5e5c9b78d04e/model.py#L185), `log p(x)` (`bits_x`) is also divided by the number of subpixels.

But when I tried to divide the MNIST loss by `image_size=28`, there was instability at around 10-15K steps:
![image](https://user-images.githubusercontent.com/2038751/72155881-c9c1a200-33bc-11ea-841c-8443a5908dc6.png)
![image](https://user-images.githubusercontent.com/2038751/72161700-46f31400-33c9-11ea-9d12-5eb20956fd6e.png)
And they encountered NaN in step 8K and 4K respectively. The NaN first pops up in `gradient/ChainLayer/Scale3/ChainLayer/Step5/ChainLayer//AffineCouplingLayer/Step5/Conv/biases_0` which is really strange because it doesn't look abnormal:
![image](https://user-images.githubusercontent.com/2038751/72161929-aea95f00-33c9-11ea-8670-864e04845e34.png)


This was with `lr=0.005` which is maybe too high as @geosada mentioned and MNIST did not have a warm-up strategy. I will try using the learning rate warm-up.

---

Also another unrelated thought--I wonder if act norm, if it acts like batch norm, will [interact with l2 regularization to produce adaptive learning rate](https:bella.cc/zettels/l2_with_batch_norm/).

**Comment by kmkolasinski** [comment]
> but in `Celeba64x64_22steps.ipynb` the learning rate schedule goes up, down, down, up---just wondering, why that schedule?

Hi, I believe I was playing with LR rate manually, so there is no specific reason for this schedule. When experimenting in jupyter notebooks, I usually start from small LR e.g. 0.0001 and test model for few epochs to check whether it trains or not. If so, I increase LR and then systematically decrease e.g. [0.001, 0.0005, 0.0001, 0.00005]. This time I increased LR at the end, probably because I noted that model is not learning fast enough, so I wanted to help it, or maybe this is just a typo. Sorry for the inconvenience.

**Comment by kmkolasinski** [comment]
Wow, you're doing a great detective work :)
```
And they encountered NaN in step 8K and 4K respectively. The NaN first pops up in gradient/ChainLayer/Scale3/ChainLayer/Step5/ChainLayer//AffineCouplingLayer/Step5/Conv/biases_0 which is really strange because it doesn't look abnormal:
```
I believe you should work with smaller LR. NFs are very complicated networks with many parts which are very fragile and they don't train fast (at least from my experience). If you have a bad luck you can always sample some hard minibatch which will generate large loss and which will break whole training. You can try to overcome this problem with some gradient clipping techniques. Or maybe one could write some custom optimizer which would reject these updates for which loss is far from the current running mean.
